paasta_tools/autoscaling_lib.py

```diff
@@ -569,104 +589,165 @@ def sort_slaves_to_kill(mesos_state, pool='default'):
-        start = int(time.time() + CLUSTER_DRAIN_TIMEOUT)
-        # Set the duration to an hour, if we haven't cleaned up and termintated by then
-        # mesos should put the slave back into the pool
-        duration = 600
-        log.info("Draining {0}".format(instances_to_kill))
-        log.info("Decreasing spot fleet capacity to: {0}".format(target_capacity))
-        if not dry_run:
-            # sort to make testing easier
-            drain([instance['ip'] for instance in sorted(instances_to_kill.values())], start, duration)
-            ec2_client.modify_spot_fleet_request(SpotFleetRequestId=sfr_id, TargetCapacity=target_capacity,
-                                                 ExcessCapacityTerminationPolicy='noTermination')
-        log.info("Waiting for instances to drain before we terminate")
-        wait_and_terminate(instances_to_kill, dry_run)
+        sfr_sorted_slaves.reverse()
+        while True:
+            if len(sfr_sorted_slaves) == 0:
+                break
+            slave_to_kill = sfr_sorted_slaves.pop()
+            instance_capacity = int(slave_to_kill['instance_weight'])
+            new_capacity = current_capacity - instance_capacity
+            if current_capacity - instance_capacity < target_capacity:
+                log.info("Terminating instance {0} with weight {1} would take us below our target of {2}, so this is as"
+                         " close to our target as we can get".format(slave_to_kill['instance_id'],
+                                                                     slave_to_kill['instance_weight'],
+                                                                     target_capacity))
+                break
+            # The start time of the maintenance window is the point at which
+            # we giveup waiting for the instance to drain and mark it for termination anyway
+            start = int(time.time() + CLUSTER_DRAIN_TIMEOUT)
+            # Set the duration to an hour, if we haven't cleaned up and termintated by then
+            # mesos should put the slave back into the pool
+            duration = 600
+            log.info("Draining {0}".format(slave_to_kill['pid']))
+            if not dry_run:
+                drain([slave_to_kill['ip']], start, duration)
+            log.info("Decreasing spot fleet capacity from {0} to: {1}".format(current_capacity, new_capacity))
+            set_spot_fleet_request_capacity(sfr_id, new_capacity, dry_run)
+            log.info("Waiting for instance to drain before we terminate")
+            try:
+                wait_and_terminate(slave_to_kill, dry_run)
+            except ClientError as e:
+                log.error("Failure when terminating: {0}: {1}".format(slave['pid'], e))
+                log.error("Setting spot fleet capacity back to {0}".format(current_capacity))
+                set_spot_fleet_request_capacity(sfr_id, current_capacity, dry_run)
```

Comment:
Yep, exactly. We set the termination policy as `NoTermination`, so reducing the capacity doesn't trigger anything on the AWS side.

I'll try and add some better error handling in this loop, I think you're right that it could be a bit more robust at checking responses.